File: client/posts/render/body.ts
import { config, boards, boardConfig, posts } from '../../state'
import { renderPostLink, renderTempLink } from './etc'
import { PostData, PostLink, TextState, commandType } from '../../common'
import { escape, makeAttrs } from '../../util'
import { parseEmbeds } from "../embed"
import highlightSyntax from "./code"

// URLs supported for linkification
const urlPrefixes = {
    'h': "http",
    'm': "magnet:?",
    'f': "ftp",
    'b': "bitcoin",
}

// Render the text body of a post
export default function renderBody(data: PostData): string {
    const state: TextState = data.state = {
        spoiler: false,
        quote: false,
        code: false,
        bold: false,
        italic: false,
        red: false,
        blue: false,
        haveSyncwatch: false,
        successive_newlines: 0,
        iDice: 0,
    }
    let html = ""

    const fn = data.editing ? parseOpenLine : parseTerminatedLine
    for (let l of data.body.split("\n")) {
        state.quote = false

        // Prevent successive empty lines
        if (html && state.successive_newlines < 2) {
            html += "<br>"
        }
        if (!l.length) {
            state.successive_newlines++
            continue
        }

        state.successive_newlines = 0
        if (l[0] === ">") {
            state.quote = true
            html += "<em>"
        }
        if (state.spoiler) {
            html += "<del>"
        }
        if (state.bold) {
            html += "<b>"
        }
        if (state.italic) {
            html += "<i>"
        }
        if (state.red) {
            html += "<span class=\"red\">"
        }
        if (state.blue) {
            html += "<span class=\"blue\">"
        }

        html += fn(l, data)

        // Close any unclosed tags
        if (state.blue) {
            html += "</span>"
        }
        if (state.red) {
            html += "</span>"
        }
        if (state.italic) {
            html += "</i>"
        }
        if (state.bold) {
            html += "</b>"
        }
        if (state.spoiler) {
            html += "</del>"
        }
        if (state.quote) {
            html += "</em>"
        }
    }

    return html
}

// Open and close any tags up to level, if they are set.
// Increment level by 1 for each tag deeper you go.
function wrapTags(level: number, state: TextState): string {
    const states = [
        state.spoiler,
        state.bold,
        state.italic,
        state.red,
        state.blue,
    ]
    const opening = [
        "<del>",
        "<b>",
        "<i>",
        "<span class=\"red\">",
        "<span class=\"blue\">",
    ]
    const closing = [
        "</del>",
        "</b>",
        "</i>",
        "</span>",
        "</span>",
    ]

    let html = ""
    for (let i = states.length - 1; i >= level; i--) {
        if (states[i]) {
            html += closing[i]
        }
    }
    if (!states[level]) {
        html += opening[level]
    }
    for (let i = level + 1; i < states.length; i++) {
        if (states[i]) {
            html += opening[i]
        }
    }
    return html
}

// Parse a single line, that is no longer being edited
function parseTerminatedLine(line: string, data: PostData): string {
    return parseCode(line, data.state, frag =>
        parseFragment(frag, data))
}

// Detect code tags
function parseCode(
    frag: string,
    state: TextState,
    fn: (frag: string) => string,
): string {
    let html = ""
    while (true) {
        const i = frag.indexOf("``")
        if (i !== -1) {
            html += formatCode(frag.slice(0, i), state, fn)
            frag = frag.substring(i + 2)
            state.code = !state.code
        } else {
            html += formatCode(frag, state, fn)
            break
        }
    }
    return html
}

function formatCode(
    frag: string,
    state: TextState,
    fn: (frag: string) => string,
): string {
    let html = ""
    if (state.code) {
        // Strip quotes
        while (frag[0] === '>') {
            html += "&gt;"
            frag = frag.slice(1)
        }
        html += highlightSyntax(frag)
    } else {
        html += parseSpoilers(frag, state, fn)
    }
    return html
}

// Inject spoiler tags and call fn on the remaining parts
function parseSpoilers(
    frag: string,
    state: TextState,
    fn: (frag: string) => string,
): string {
    const _fn = (frag: string) =>
        parseBolds(frag, state, fn)
    let html = ""
    while (true) {
        const i = frag.indexOf("**")
        if (i !== -1) {
            html += _fn(frag.slice(0, i)) + wrapTags(0, state)
            state.spoiler = !state.spoiler
            frag = frag.substring(i + 2)
        } else {
            html += _fn(frag)
            break
        }
    }
    return html
}

// Inject bold tags and call fn on the remaining parts
function parseBolds(
    frag: string,
    state: TextState,
    fn: (frag: string) => string,
): string {
    const _fn = (frag: string) =>
        parseItalics(frag, state, fn)
    let html = ""
    while (true) {
        const i = frag.indexOf("@@")
        if (i !== -1) {
            html += _fn(frag.slice(0, i)) + wrapTags(1, state)
            state.bold = !state.bold
            frag = frag.substring(i + 2)
        } else {
            html += _fn(frag)
            break
        }
    }
    return html
}

// Inject italic tags and call fn on the remaining parts
function parseItalics(
    frag: string,
    state: TextState,
    fn: (frag: string) => string,
): string {
    const _fn = (frag: string) => parseReds(frag, state, fn)
    let html = ""
    while (true) {
        const i = frag.indexOf("~~")
        if (i !== -1) {
            html += _fn(frag.slice(0, i)) + wrapTags(2, state)
            state.italic = !state.italic
            frag = frag.substring(i + 2)
        } else {
            html += _fn(frag)
            break
        }
    }
    return html
}

// Inject red color tags and call fn on the remaining parts
function parseReds(
    frag: string,
    state: TextState,
    fn: (frag: string) => string,
): string {
    const _fn = (frag: string) =>
        parseBlues(frag, state, fn)
    const _rbText = boardConfig.rbText ? () => {
        const wrapped = wrapTags(3, state)
        state.red = !state.red
        return wrapped
    } : () => ""
    let html = ""

    while (true) {
        const i = frag.indexOf("^r")
        if (i !== -1) {
            html += _fn(frag.slice(0, i)) + _rbText()
            frag = frag.substring(i + 2)
        } else {
            html += _fn(frag)
            break
        }
    }

    return html
}

// Inject blue color tags and call fn on the remaining parts
function parseBlues(
    frag: string,
    state: TextState,
    fn: (frag: string) => string,
): string {
    const _rbText = boardConfig.rbText ? () => {
        const wrapped = wrapTags(4, state)
        state.blue = !state.blue
        return wrapped
    } : () => ""
    let html = ""

    while (true) {
        const i = frag.indexOf("^b")
        if (i !== -1) {
            html += fn(frag.slice(0, i)) + _rbText()
            frag = frag.substring(i + 2)
        } else {
            html += fn(frag)
            break
        }
    }

    return html
}

// Parse a line that is still being edited
function parseOpenLine(line: string, { state }: PostData): string {
    return parseCode(line, state, parseOpenLinks)
}

// Parse temporary links, that still may be edited
function parseOpenLinks(frag: string): string {
    let html = ""
    const words = frag.split(" ")
    for (let i = 0; i < words.length; i++) {
        if (i !== 0) {
            html += " "
        }

        // Split leading and trailing punctuation, if any
        const [leadPunct, word, trailPunct] = splitPunctuation(words[i])
        if (leadPunct) {
            html += leadPunct
        }

        let matched = false
        if (word && word[0] === ">") {
            const m = word.match(/^>>(>*)(\d+)$/)
            if (m) {
                const id = parseInt(m[2])
                if (posts.has(id)) {
                    html += m[1] + renderTempLink(id)
                    matched = true
                }
            }
        }
        if (!matched) {
            html += escape(word)
        }
        if (trailPunct) {
            html += trailPunct
        }
    }
    return html
}

// Parse a line fragment
function parseFragment(frag: string, data: PostData): string {
    let html = ""
    const words = frag.split(" ")
    for (let i = 0; i < words.length; i++) {
        if (i !== 0) {
            html += " "
        }

        // Split leading and trailing punctuation, if any
        let [leadPunct, word, trailPunct] = splitPunctuation(words[i])
        if (leadPunct) {
            html += leadPunct
        }
        if (!word) {
            if (trailPunct) {
                html += trailPunct
            }
            continue
        }

        let m: RegExpMatchArray,
            matched = false
        switch (word[0]) {
            case ">":
                // Post links
                m = word.match(/^>>(>*)(\d+)$/)
                if (m) {
                    html += parsePostLink(m, data.links)
                    matched = true
                    break
                }

                // Internal and custom reference URLs
                m = word.match(/^>>>(>*)\/(\w+)\/$/)
                if (m) {
                    html += parseReference(m)
                    matched = true
                }
                break
            case "#": // Hash commands
                if (data.state.quote) {
                    break
                }
                m = word.match(/^#(flip|\d*d\d+|8ball|pyu|pcount|sw(?:\d+:)?\d+:\d+(?:[+-]\d+)?|roulette|rcount)$/)
                if (m) {
                    html += parseCommand(m[1], data)
                    matched = true
                    break
                }
                break
            default:
                // Generic HTTP(S) URLs, magnet links and embeds
                // Checking the first byte is much cheaper than a function call.
                // Do that first, as most cases won't match.
                const pre = urlPrefixes[word[0]]
                if (pre && word.startsWith(pre)) {
                    html += parseURL(word)
                    matched = true
                    break
                }
        }

        if (!matched) {
            html += escape(word)
        }
        if (trailPunct) {
            html += trailPunct
        }
    }

    return html
}

// Verify and render a link to other posts
function parsePostLink(m: string[], links: PostLink[]): string {
    if (!links) {
        return m[0]
    }
    const id = parseInt(m[2])
    let data: PostLink
    for (let l of links) {
        if (l.id === id) {
            data = l
            break
        }
    }
    if (!data) {
        return m[0]
    }
    return m[1] + renderPostLink(data)
}

// Parse internal or customly set reference URL
function parseReference(m: string[]): string {
    let href: string
    if (boards.includes(m[2])) {
        href = `/${m[2]}/`
    } else if (m[2] in config.links) {
        href = config.links[m[2]]
    } else {
        return m[0]
    }
    return m[1] + newTabLink(href, `>>>/${m[2]}/`)
}

// Render and anchor link that opens in a new tab
function newTabLink(href: string, text: string): string {
    const attrs = {
        rel: "noreferrer",
        href: escape(href),
        target: "_blank",
    }
    return `<a ${makeAttrs(attrs)}>${escape(text)}</a>`
}

// Parse generic URLs and embed, if applicable
function parseURL(bit: string): string {
    const embed = parseEmbeds(bit)
    if (embed) {
        return embed
    }

    try {
        new URL(bit) // Will throw, if invalid URL
        if (bit[0] == "m") { // Don't open a new tab for magnet links
            bit = escape(bit)
            return bit.link(bit)
        }
        return newTabLink(bit, bit)
    } catch (e) {
        return escape(bit)
    }
}

// Parse a hash command
function parseCommand(bit: string, { commands, state }: PostData): string {
    // Guard against invalid dice rolls
    if (!commands || !commands[state.iDice]) {
        return "#" + bit
    }

    // Protect from index shifts on boardConfig.pyu toggle
    if (!boardConfig.pyu) {
        switch (commands[state.iDice].type) {
            case commandType.pyu:
            case commandType.pcount:
                state.iDice++
                return "#" + bit
        }
    }

    let formatting = "<strong>"
    let inner: string
    let indirect = false;
    switch (bit) {
        case "flip":
            inner = commands[state.iDice++].val ? "flap" : "flop"
            break
        case "8ball":
            inner = escape(commands[state.iDice++].val.toString())
            break
        case "pyu":
        case "pcount":
            // Protect from index shifts on boardConfig.pyu toggle
            if (!boardConfig.pyu) {
                break
            }
        case "rcount":
            switch (commands[state.iDice].type) {
                case commandType.pyu:
                case commandType.pcount:
                case commandType.rcount:
                    inner = commands[state.iDice++].val.toString()
            }

            break
        case "roulette":
            let val = commands[state.iDice++].val
            inner = val[0].toString() + "/" + val[1].toString()
            // set formatting if the poster died
            if (val[0] == 1) {
                formatting = "<strong class=\"dead\">"
            }
            break
        default:
            indirect = true;
            if (bit.startsWith("sw")) {
                // Protect from various index shift attacks due to dynamic typing
                if (commands[state.iDice + 1].type !== commandType.syncWatch) {
                    return "#" + bit;
                }
                return formatSyncwatch(bit, commands[state.iDice++].val, state)
            }

            // Validate dice
            if (commands[state.iDice + 1].type !== commandType.dice) {
                return "#" + bit;
            }
            const m = bit.match(/^(\d*)d(\d+)$/)
            if (parseInt(m[1]) > 10 || parseInt(m[2]) > 10000) {
                return "#" + bit
            }
            const sides = parseInt(m[2])

            const rolls = commands[state.iDice++].val as number[]
            inner = ""
            let sum = 0
            for (let i = 0; i < rolls.length; i++) {
                if (i) {
                    inner += " + "
                }
                sum += rolls[i]
                inner += rolls[i]
            }
            if (rolls.length > 1) {
                inner += " = " + sum
            }

            formatting = getRollFormatting(rolls.length, sides, sum)
    }

    // Protect from various index shift attacks due to dynamic typing
    const commandMatchers: { [bit: string]: commandType } = {
        flip: commandType.flip,
        "8ball": commandType.eightBall,
        pyu: commandType.pyu,
        pcount: commandType.pcount,
        rcount: commandType.rcount,
        roulette: commandType.roulette,
    }
    if (indirect && commandMatchers[bit] !== commands[state.iDice - 1].type) {
        return "#" + bit;
    }

    return `${formatting}#${bit} (${inner})</strong>`
}

function getRollFormatting(numberOfDice: number, facesPerDie: number, sum: number): string {
    const maxRoll = numberOfDice * facesPerDie
    // no special formatting for small rolls
    if (maxRoll < 10 || facesPerDie == 1) {
        return "<strong>"
    }

    if (maxRoll == sum) {
        return "<strong class=\"super_roll\">";
    } else if (sum == numberOfDice) {
        return "<strong class=\"kuso_roll\">";
    } else if (sum == 69 || sum == 6969) {
        return "<strong class=\"lewd_roll\">";
    } else if (checkEm(sum)) {
        if (sum < 100) {
            return "<strong class=\"dubs_roll\">";
        } else if (sum < 1000) {
            return "<strong class=\"trips_roll\">";
        } else if (sum < 10000) {
            return "<strong class=\"quads_roll\">";
        } else {// QUINTS!!!
            return "<strong class=\"rainbow_roll\">";
        }
    }
    return "<strong>"
}

// If num is made of the same digit repeating
function checkEm(num: number): boolean {
    if (num < 10) {
        return false
    }
    const digit = num % 10
    while (true) {
        num = Math.floor(num / 10)
        if (num == 0) {
            return true
        }
        if (num % 10 != digit) {
            return false
        }
    }
}

// Format a synchronized time counter
function formatSyncwatch(bit: string, val: number[], state: TextState): string {
    state.haveSyncwatch = true
    const attrs = {
        class: "embed syncwatch",
        "data-hour": val[0].toString(),
        "data-min": val[1].toString(),
        "data-sec": val[2].toString(),
        "data-start": val[3].toString(),
        "data-end": val[4].toString()
    }
    return `<em><strong ${makeAttrs(attrs)}>syncwatch</strong></em>`
}

// Splits off one byte of leading and trailing punctuation, if any, and returns
// the 3 split parts. If there is no edge punctuation, the respective string
// is empty.
function splitPunctuation(word: string): [string, string, string] {
    const re: [string, string, string] = ["", word, ""]
    re[1] = word

    // Split leading
    if (re[1].length < 2) {
        return re
    }
    if (isPunctuation(re[1][0])) {
        re[0] = re[1][0]
        re[1] = re[1].slice(1)
    }

    // Split trailing
    const l = re[1].length
    if (l < 2) {
        return re
    }
    if (isPunctuation(re[1][l - 1])) {
        re[2] = re[1][l - 1]
        re[1] = re[1].slice(0, -1)
    }

    return re
}

// Return if b is a punctuation byte
function isPunctuation(b: string): boolean {
    switch (b) {
        case '!':
        case '"':
        case '\'':
        case '(':
        case ')':
        case ',':
        case '-':
        case '.':
        case ':':
        case ';':
        case '?':
        case '[':
        case ']':
            return true
        default:
            return false
    }
}

File: go/src/meguca/templates/body.go
package templates

import (
	"fmt"
	"html"
	"meguca/common"
	"meguca/config"
	"meguca/util"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"unicode"

	"github.com/valyala/quicktemplate"
)

// Embeddable URL types
const (
	youTube = iota
	soundCloud
	vimeo
)

var (
	linkRegexp      = regexp.MustCompile(`^>>(>*)(\d+)$`)
	referenceRegexp = regexp.MustCompile(`^>>>(>*)\/(\w+)\/$`)
	dickRegex       = regexp.MustCompile(`(?i)(dick|cock)s?`)

	providers = map[int]string{
		youTube:    "Youtube",
		soundCloud: "SoundCloud",
		vimeo:      "Vimeo",
	}
	embedPatterns = [...]struct {
		typ  int
		patt *regexp.Regexp
	}{
		{
			youTube,
			regexp.MustCompile(`https?:\/\/(?:[^\.]+\.)?youtube\.com\/watch\/?\?(?:.+&)?v=([^&]+)`),
		},
		{
			youTube,
			regexp.MustCompile(`https?:\/\/(?:[^\.]+\.)?(?:youtu\.be|youtube\.com\/embed)\/([a-zA-Z0-9_-]+)`),
		},
		{
			soundCloud,
			regexp.MustCompile(`https?:\/\/soundcloud.com\/.*`),
		},
		{
			vimeo,
			regexp.MustCompile(`https?:\/\/(?:www\.)?vimeo\.com\/.+`),
		},
	}

	// URLs supported for linkification
	urlPrefixes = map[byte]string{
		'h': "http",
		'm': "magnet:?",
		'i': "irc",
		'f': "ftp",
		'b': "bitcoin",
	}
)

type bodyContext struct {
	index bool     // Rendered for an index page
	state struct { // Body parser state
		spoiler, quote, code bool
		newlines             uint
		iDice                int
	}
	common.Post
	OP    uint64
	board string
	quicktemplate.Writer
}

// Render the text body of a post
func streambody(
	w *quicktemplate.Writer,
	p common.Post,
	op uint64,
	board string,
	index bool,
) {
	c := bodyContext{
		index:  index,
		Post:   p,
		OP:     op,
		board:  board,
		Writer: *w,
	}

	var fn func(string)
	if c.Editing {
		fn = c.parseOpenLine
	} else {
		fn = c.parseTerminatedLine
	}

	for i, l := range strings.Split(c.Body, "\n") {
		c.state.quote = false

		// Prevent successive empty lines
		if i != 0 && c.state.newlines < 2 {
			c.string("<br>")
		}
		if len(l) == 0 {
			c.state.newlines++
			continue
		}

		c.state.newlines = 0
		if l[0] == '>' {
			c.string("<em>")
			c.state.quote = true
		}
		if c.state.spoiler {
			c.string("<del>")
		}

		fn(l)

		if c.state.spoiler {
			c.string("</del>")
		}
		if c.state.quote {
			c.string("</em>")
		}
	}
}

// Write string without escaping
func (c *bodyContext) string(s string) {
	c.N().S(s)
}

// Escape and write string
func (c *bodyContext) escape(s string) {
	c.E().S(s)
}

// Write a byte without heap allocations or escaping
func (c *bodyContext) byte(b byte) {
	buf := [1]byte{b}
	c.N().SZ(buf[:])
}

// Parse a line that is no longer being edited
func (c *bodyContext) parseTerminatedLine(line string) {
	c.parseCode(line, (*c).parseFragment)
}

// Detect code tags
func (c *bodyContext) parseCode(frag string, fn func(string)) {
	for {
		i := strings.Index(frag, "``")
		if i != -1 {
			c.formatCode(frag[:i], fn)
			frag = frag[i+2:]
			c.state.code = !c.state.code
		} else {
			c.formatCode(frag, fn)
			break
		}
	}
}

func (c *bodyContext) formatCode(frag string, fn func(string)) {
	if c.state.code {
		// Strip quotes
		for len(frag) != 0 && frag[0] == '>' {
			c.string(`&gt;`)
			frag = frag[1:]
		}
		c.N().Z(highlightSyntax(frag))
	} else {
		c.parseSpoilers(frag, fn)
	}
}

// Injects spoiler tags and calls fn on the remaining parts
func (c *bodyContext) parseSpoilers(frag string, fn func(string)) {
	for {
		i := strings.Index(frag, "**")
		if i != -1 {
			fn(frag[:i])
			if c.state.quote {
				c.string("</em>")
			}
			if c.state.spoiler {
				c.string("</del>")
			} else {
				c.string("<del>")
			}
			if c.state.quote {
				c.string("<em>")
			}

			c.state.spoiler = !c.state.spoiler
			frag = frag[i+2:]
		} else {
			fn(frag)
			break
		}
	}
}

// Parse a line fragment
func (c *bodyContext) parseFragment(frag string) {
	// Leading and trailing punctuation, if any
	var leadPunct, trailPunct byte

	for i, word := range strings.Split(frag, " ") {
		if i != 0 {
			c.byte(' ')
		}

		// Strip leading and trailing punctuation and commit separately
		leadPunct, word, trailPunct = util.SplitPunctuationString(word)
		if leadPunct != 0 {
			c.byte(leadPunct)
		}

		if len(word) == 0 {
			goto end
		}
		switch word[0] {
		case '>': // Links
			if m := linkRegexp.FindStringSubmatch(word); m != nil {
				// Post links
				c.parsePostLink(m)
				goto end
			} else if m := referenceRegexp.FindStringSubmatch(word); m != nil {
				// Internal and custom reference URLs
				c.parseReference(m)
				goto end
			}
		case '#': // Hash commands
			if m := common.CommandRegexp.FindStringSubmatch(word); m != nil {
				c.parseCommands(string(m[1]))
				goto end
			}
		default: // Generic HTTP(S) URLs and magnet links
			// Checking the first byte is much cheaper than a function call. Do
			// that first, as most cases won't match.
			pre, ok := urlPrefixes[word[0]]
			if ok && strings.HasPrefix(word, pre) {
				c.parseURL(word)
				goto end
			}
		}

		if c.board == "a" && dickRegex.MatchString(word) {
			if unicode.IsUpper(rune(word[0])) {
				word = "Privilege"
			} else {
				word = "privilege"
			}
		}
		c.escape(word)

	end:
		// Write trailing punctuation, if any
		if trailPunct != 0 {
			c.byte(trailPunct)
		}
	}
}

// Parse a potential link to a post
func (c *bodyContext) parsePostLink(m []string) {
	if c.Links == nil {
		c.string(m[0])
		return
	}

	id, _ := strconv.ParseUint(string(m[2]), 10, 64)
	var op uint64
	for _, l := range c.Links {
		if l[0] == id {
			op = l[1]
			break
		}
	}
	if op == 0 {
		c.string(m[0])
		return
	}

	if len(m[1]) != 0 { // Write extra quotes
		c.string(m[1])
	}
	streampostLink(&c.Writer, id, op != c.OP, c.index)
}

// Parse internal or customly set reference URL
func (c *bodyContext) parseReference(m []string) {
	var (
		m2   = string(m[2])
		href string
	)
	if config.IsBoard(m2) {
		href = fmt.Sprintf("/%s/", m2)
	} else if href = config.Get().Links[m2]; href != "" {
	} else {
		c.string(m[0])
		return
	}

	if len(m[1]) != 0 {
		c.string(m[1])
	}
	c.newTabLink(href, fmt.Sprintf(">>>/%s/", string(m[2])))
}

// Format and anchor link that opens in a new tab
func (c *bodyContext) newTabLink(href, text string) {
	c.string(`<a rel="noreferrer" href="`)
	c.escape(href)
	c.string(`" target="_blank">`)
	c.escape(text)
	c.string(`</a>`)
}

// Parse generic URLs and magnet links
func (c *bodyContext) parseURL(bit string) {
	s := string(bit)
	u, err := url.Parse(s)
	switch {
	case err != nil || u.Path == s: // Invalid or empty path
		c.escape(bit)
	case c.parseEmbeds(bit):
	case bit[0] == 'm': // Don't open a new tab for magnet links
		s = html.EscapeString(s)
		c.string(`<a rel="noreferrer" href="`)
		c.string(s)
		c.string(`">`)
		c.string(s)
		c.string(`</a>`)
	default:
		c.newTabLink(s, s)
	}
}

// Parse select embeddable URLs. Returns, if any found.
func (c *bodyContext) parseEmbeds(s string) bool {
	for _, t := range embedPatterns {
		if !t.patt.MatchString(s) {
			continue
		}

		c.string(`<em><a rel="noreferrer" class="embed" target="_blank" data-type="`)
		c.N().D(t.typ)
		c.string(`" href="`)
		c.escape(s)
		c.string(`">[`)
		c.string(providers[t.typ])
		c.string(`] ???</a></em>`)

		return true
	}
	return false
}

// Parse a hash command
func (c *bodyContext) parseCommands(bit string) {
	// Guard against invalid dice rolls
	if c.Commands == nil || c.state.iDice > len(c.Commands)-1 {
		c.writeInvalidCommand(bit)
		return
	}

	inner := make([]byte, 0, 32)
	val := c.Commands[c.state.iDice]
	switch bit {
	case "flip":
		var s string
		if val.Flip {
			s = "flap"
		} else {
			s = "flop"
		}
		inner = append(inner, s...)
		c.state.iDice++
	case "8ball":
		inner = append(inner, val.Eightball...)
		c.state.iDice++
	case "pyu", "pcount":
		inner = strconv.AppendUint(inner, val.Pyu, 10)
		c.state.iDice++
	default:
		if strings.HasPrefix(bit, "sw") {
			c.formatSyncwatch(val.SyncWatch)
			c.state.iDice++
			return
		}

		// Validate dice
		m := common.DiceRegexp.FindStringSubmatch(bit)
		if m[1] != "" {
			if rolls, err := strconv.Atoi(m[1]); err != nil || rolls > 10 {
				c.writeInvalidCommand(bit)
				return
			}
		}
		sides, err := strconv.Atoi(m[2])
		if err != nil || sides > common.MaxDiceSides {
			c.writeInvalidCommand(bit)
			return
		}

		c.state.iDice++
		var sum uint64
		for i, roll := range val.Dice {
			if i != 0 {
				inner = append(inner, " + "...)
			}
			sum += uint64(roll)
			inner = strconv.AppendUint(inner, uint64(roll), 10)
		}
		if len(val.Dice) > 1 {
			inner = append(inner, " = "...)
			inner = strconv.AppendUint(inner, sum, 10)
		}
	}

	c.string(`<strong>#`)
	c.string(bit)
	c.string(` (`)
	c.N().Z(inner)
	c.string(`)</strong>`)
}

// Format a synchronized time counter
func (c *bodyContext) formatSyncwatch(val [5]uint64) {
	c.string(`<em><strong class="embed syncwatch" data-hour=`)
	c.uint64(val[0])
	c.string(` data-min=`)
	c.uint64(val[1])
	c.string(` data-sec=`)
	c.uint64(val[2])
	c.string(` data-start=`)
	c.uint64(val[3])
	c.string(` data-end=`)
	c.uint64(val[4])
	c.string(`>syncwatch</strong></em>`)
}

func (c *bodyContext) uint64(i uint64) {
	c.string(strconv.FormatUint(i, 10))
}

// If command validation failed, simply write the string
func (c *bodyContext) writeInvalidCommand(s string) {
	c.byte('#')
	c.escape(s)
}

// Parse a line that is still being edited
func (c *bodyContext) parseOpenLine(line string) {
	c.parseCode(line, func(s string) {
		c.escape(s)
	})
}
